File: packages/core/src/api/graphOrder.ts
import { BlockCategory } from '../model/BlockCategory.js';
import type { FlowDefinition } from '../model/FlowDefinition.js';
import type { FlowGraph } from '../model/FlowGraph.js';
import type { FlowNode } from '../model/FlowNode.js';
import { getBlock } from './BlockRegistry.js';
import { getFlows } from './normalizeGraph.js';

/** Topological walk from the single SOURCE node within one flow. */
export function orderedNodesFromFlow(flow: FlowDefinition): FlowNode[] {
  const sourceNode = flow.nodes.find(
    (n) => getBlock(n.blockType)?.category === BlockCategory.SOURCE
  );
  if (!sourceNode) return [];

  const adjMap = new Map<string, string[]>();
  for (const n of flow.nodes) adjMap.set(n.id, []);
  for (const e of flow.edges) adjMap.get(e.source)?.push(e.target);

  const nodeMap = new Map(flow.nodes.map((n) => [n.id, n]));
  const ordered: FlowNode[] = [];
  const visited = new Set<string>();

  function dfs(id: string) {
    if (visited.has(id)) return;
    visited.add(id);
    const node = nodeMap.get(id);
    if (node) ordered.push(node);
    for (const child of adjMap.get(id) ?? []) dfs(child);
  }

  dfs(sourceNode.id);
  return ordered;
}

/** @deprecated Use orderedNodesFromFlow with active flow */
export function orderedNodesFromGraph(graph: FlowGraph): FlowNode[] {
  const flows = getFlows(graph);
  if (flows.length === 0) return [];
  return orderedNodesFromFlow(flows[0]!);
}

File: java/flowcamel-core/src/main/java/com/flowcamel/core/graph/GraphOrder.java
package com.flowcamel.core.graph;

import com.flowcamel.core.model.BlockCategory;
import com.flowcamel.core.model.FlowDefinition;
import com.flowcamel.core.model.FlowGraph;
import com.flowcamel.core.model.FlowNode;
import com.flowcamel.core.registry.BlockRegistry;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class GraphOrder {
  private GraphOrder() {}

  public static List<FlowNode> orderedNodesFromFlow(FlowDefinition flow) {
    FlowNode source =
        flow.nodes.stream()
            .filter(
                n ->
                    BlockRegistry.getBlock(n.blockType)
                        .map(b -> b.category == BlockCategory.SOURCE)
                        .orElse(false))
            .findFirst()
            .orElse(null);
    if (source == null) return List.of();

    Map<String, List<String>> adj = new HashMap<>();
    for (FlowNode n : flow.nodes) adj.put(n.id, new ArrayList<>());
    for (var e : flow.edges) adj.computeIfAbsent(e.source, k -> new ArrayList<>()).add(e.target);

    Map<String, FlowNode> nodeMap = new HashMap<>();
    for (FlowNode n : flow.nodes) nodeMap.put(n.id, n);

    List<FlowNode> ordered = new ArrayList<>();
    Set<String> visited = new HashSet<>();
    dfs(source.id, adj, nodeMap, visited, ordered);
    return ordered;
  }

  /** @deprecated use {@link #orderedNodesFromFlow} */
  public static List<FlowNode> orderedNodesFromGraph(FlowGraph graph) {
    List<FlowDefinition> flows = GraphNormalizer.getFlows(graph);
    if (flows.isEmpty()) return List.of();
    return orderedNodesFromFlow(flows.getFirst());
  }

  private static void dfs(
      String id,
      Map<String, List<String>> adj,
      Map<String, FlowNode> nodeMap,
      Set<String> visited,
      List<FlowNode> ordered) {
    if (visited.contains(id)) return;
    visited.add(id);
    FlowNode node = nodeMap.get(id);
    if (node != null) ordered.add(node);
    for (String child : adj.getOrDefault(id, List.of())) dfs(child, adj, nodeMap, visited, ordered);
  }
}
